tests/training_success/explore_schema.py

```diff
@@ -0,0 +1,422 @@
+        cfg = merge_dict(child_config_options, config_options)
+        dq.append((cfg, visited))
+    return dq
+
+
+def explore_from_all_of(
+    config_options: Dict[str, Any], item: Dict[str, Any], key_so_far: str
+) -> Deque[Tuple[Dict, bool]]:
+    """Takes a child of `allOf` and calls `explore_properties` on it."""
+    for key in item["if"]["properties"]:
+        config_options[key_so_far + "." + key] = item["if"]["properties"][key]["const"]
+    properties = item["then"]["properties"]
+    raw_entry = deque([(copy.deepcopy(config_options), False)])
+    return explore_properties(properties, parent_key=key_so_far, dq=raw_entry)
+
+
+def get_potential_values(item: Dict[str, Any]) -> List[Union[ParameterBaseTypes, List[ParameterBaseTypes]]]:
+    """Returns a list of values to explore for a config parameter.
+
+    Param:
+        item: config parameter-specific dictionary. Considered as a leaf in the schema. Contains type, default, and
+            parameter metadata, etc.
+    """
+    temp = []
+    # Case where we're using OneOf (e.g. to allow batch size 'auto' and integers)
+    if isinstance(item["type"], list):
+        for property_type in item["type"]:
+            temp += handle_property_type(property_type, item)
+    else:
+        temp += handle_property_type(item["type"], item)
+
+    # Make sure values are unique. Not using set because some values are unhashable.
+    unique_temp = []
+    for temp_item in temp:
+        if temp_item not in unique_temp:
+            unique_temp.append(temp_item)
+    return unique_temp
+
+
+def handle_property_type(
+    property_type: str, item: Dict[str, Any]
+) -> List[Union[ParameterBaseTypes, List[ParameterBaseTypes]]]:
+    """Return possible parameter values for a parameter type.
+
+    Args:
+        property_type: type of the parameter (e.g. array, number, etc.)
+        item: dictionary containing details on the parameter such as default, min and max values.
+    """
+    parameter_metadata = item.get("parameter_metadata", None)
+    if not parameter_metadata:
+        return []
+
+    # don't explore internal only parameters.
+    if parameter_metadata.get("internal_only", True):
+        return []
+
+    # don't explore parameters that have expected impact less than HIGH.
+    if parameter_metadata.get("expected_impact", ExpectedImpact.LOW) < ExpectedImpact.HIGH:
+        return []
+
+    if property_type == "number":
+        return explore_number(item)
+    elif property_type == "integer":
+        return explore_integer(item)
+    elif property_type == "string":
+        return explore_string(item)
+    elif property_type == "boolean":
+        return explore_boolean()
+    elif property_type == "null":
+        return explore_null()
+    elif property_type == "array":
+        return explore_array(item)
+    else:
+        return []
+
+
+def explore_array(item: Dict[str, Any]) -> List[List[ParameterBaseTypes]]:
```

Comment:
suggestion: Put the `explore_<atomic_type>()` functions in a separate file, i.e. `sample_leaf_property.py` or `explore_leaf_property.py`